rfcs/0034-system-parameters.md

```diff
@@ -0,0 +1,135 @@
+
+## Design
+
+```mermaid
+graph LR
+    all[Does it apply to the RisingWave instance?]
+    all ==No==> process["Process Configs (Local)"]
+    all ==Yes==> system["System Parameters (on Meta)"]
+    process --> cli_args(Command-line arguments)
+    process --> conf_file("Config file (.toml)")
+    system --> sql_cmd(SQL interface: SET / SHOW)
+    system --> cloud(Web Console)
+```
+
+### System Parameters
+
+What we need is a centralized source-of-truth of cluster configurations. Obviously, the best choice for RisingWave is the Meta Service and the underlying etcd. 
+
+Here we could borrow the syntax from Postgres:
+
+```sql
+-- Set or reset a parameter
+ALTER SYSTEM / DATABASE SET parameter TO value / 'value' / DEFAULT;
+
+-- List all parameters keys & values
+SHOW PARAMETERS;
+```
+
+> **Note**: PG uses `SHOW ALL` because it doesn't has other `SHOW` commands like us. Which one is better?
+
+
+References:
+
+- [PostgreSQL: Documentation: 15: ALTER SYSTEM](https://www.postgresql.org/docs/current/sql-altersystem.html)
+- [PostgreSQL: Documentation: 15: 20.1. Setting Parameters](https://www.postgresql.org/docs/current/config-setting.html)
+- [PostgreSQL: Documentation: 15: SHOW](https://www.postgresql.org/docs/current/sql-show.html)
+
+Similar to other system metadata e.g. catalog,
+
+- Changes to global parameters should be propagated to all nodes, in the cluster in a deterministic order ideally
+- When a node reboots or recovers, the latest snapshot of parameters should be passed from Meta Service.
+
+Therefore, the Notification Service seems to fit the scene well.
+
+### Session Parameters
+
+PostgreSQL allows users to set parameters for the current session as well.
+
+```sql
+SET parameter TO value / 'value' / DEFAULT;
+```
+
+References:
+
+- [PostgreSQL: Documentation: 15: 20.1. Setting Parameters](https://www.postgresql.org/docs/current/config-setting.html)
+
+Previously, we had already introduced [session parameters in RisingWave](https://github.com/risingwavelabs/risingwave/blob/53f7e0db772ac7e51773791bb8301624ed763ae8/src/common/src/session_config/mod.rs#L265). Some of them should be system parameters like `batch_enable_sort_agg` or `query_mode`. Postgres **enforces** that session parameters must also be system paramters, we can follow this rule as well.
+
+### Process Configs
+
+Let's rethink the current design of CLI options and config files. ([#5676](https://github.com/risingwavelabs/risingwave/issues/5676)) Given the existence of system parameters, some existing parameters may need to be moved to system parameters, like `data_directory` as mentioned before.
+
+We propose to consider CLI options as a way to override config files (i.e. in term of priority, CLI options -> config file -> defaults); as a result, the items in CLI options are a subset of the config file. The reasons for this include
```

Comment:
Shall we also consider supporting env vars as one of the config sources? In containerized deployments, it is way easier to set an env var than update a mounted file, not to mention that RisingWave is a distributed system that makes config dispatching a mess: it's hard to maintain a lot of local config files.

Reply:
> [...] not to mention that RisingWave is a distributed system that makes config dispatching a mess: it's hard to maintain a lot of local config files.

IMHO K8s ConfigMaps may help here, since dispatch and updates will happened automatically. Maybe it would be a good alternative to combine CMs with env vars? A local env var could overwrite the CM and you could thus customise the global configs for every pod.